ncm-spma/src/main/perl/spma/apt.pm

```diff
@@ -0,0 +1,433 @@
+
+=item
+
+Upgrade already installed packages
+
+=item
+
+Install packages specified in the profile that are not installed
+
+=item
+
+Mark any packages installed but not in the profile as automatically installed
+
+=item
+
+Ask apt to remove all automatically installed packages that are not satisfying dependencies of other packages
+
+=back
+
+Support for groups (tasks) is not currently provided.
+
+=head1 RESOURCES
+
+Only a very minimal schema is implemented.
+
+Repositories listed under C</software/repositories> will be configured,
+URLs should be followed by the suite and sections required e.g. http://example.org/debian unstable main
+
+Packages listed under C</software/packages> will be installed, version and architecture locking (including multiarch) is fully implemented.
+
+=cut
+
+use strict;
+use warnings;
+use NCM::Component;
+our $EC=LC::Exception::Context->new->will_store_all;
+our @ISA = qw(NCM::Component);
+use EDG::WP4::CCM::Element qw(escape unescape);
+
+use CAF::Process;
+use CAF::FileWriter;
+use CAF::FileEditor;
+use LC::Exception qw(SUCCESS);
+use Set::Scalar;
+use File::Path qw(mkpath);
+use Text::Glob qw(match_glob);
+use Readonly;
+
+Readonly::Scalar my $DIR_REPOS => "/etc/apt/sources.list.d";
+Readonly::Scalar my $DIR_PREFERENCES => "/etc/apt/preferences.d";
+Readonly::Scalar my $TEMPLATE_REPOS => "spma/apt-source.tt";
+Readonly::Scalar my $TEMPLATE_PREFERENCES => "spma/apt-preferences.tt";
+Readonly::Scalar my $TREE_REPOS => "/software/repositories";
+Readonly::Scalar my $TREE_PKGS => "/software/packages";
+Readonly::Scalar my $TREE_COMPONENT => "/software/components/${project.artifactId}";
+Readonly::Scalar my $BIN_APT_GET => "/usr/bin/apt-get";
+Readonly::Scalar my $BIN_APT_MARK => "/usr/bin/apt-mark";
+Readonly::Scalar my $BIN_DPKG_QUERY => "/usr/bin/dpkg-query";
+Readonly::Scalar my $CMD_APT_UPDATE => [$BIN_APT_GET, qw(-qq update)];
+Readonly::Scalar my $CMD_APT_UPGRADE => [$BIN_APT_GET, qw(-qq dist-upgrade)];
+Readonly::Scalar my $CMD_APT_INSTALL => [$BIN_APT_GET, qw(-qq install)];
+Readonly::Scalar my $CMD_APT_AUTOREMOVE => [$BIN_APT_GET, qw(-qq autoremove)];
+Readonly::Scalar my $CMD_APT_MARK => [$BIN_APT_MARK, qw(-qq)];
+Readonly::Scalar my $CMD_DPKG_QUERY => [$BIN_DPKG_QUERY, qw(-W -f=${db:Status-Abbrev};${Package}\n)];
+
+our $NoActionSupported = 1;
+
+
+# If user packages are not allowed, removes any repositories present
+# in the system that are not listed in $allowed_repos.
+sub cleanup_old_repos
+{
+    my ($self, $repo_dir, $allowed_repos, $allow_user_pkgs) = @_;
+    $self->debug(5, 'Entered cleanup_old_repos()');
+
+    return 1 if $allow_user_pkgs;
+
+    my $dir;
+    if (!opendir($dir, $repo_dir)) {
+        $self->error("Unable to read repositories in $repo_dir");
+        return 0;
+    }
+
+    my $current = Set::Scalar->new(map(m{(.*)\.list$}, readdir($dir)));
+
+    closedir($dir);
+    my $allowed = Set::Scalar->new(map($_->{name}, @$allowed_repos));
```

Comment:
I know standard Quattor code relies on parentheses a lot. But Perl usually prefers

```perl
map {$_->{name}}, @$allowed_repos
```
Parentheses in `map`, `grep` and `sort` cause lots of headaches because of ambiguities.